File: Kernel Simulator/Misc/Screensaver/Displays/Starfield.vb

'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Namespace Misc.Screensaver.Displays
    Public Module StarfieldSettings

        Private _starfieldDelay As Integer = 10

        ''' <summary>
        ''' [Starfield] How many milliseconds to wait before making the next write?
        ''' </summary>
        Public Property StarfieldDelay As Integer
            Get
                Return _starfieldDelay
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 10
                _starfieldDelay = value
            End Set
        End Property

    End Module

    Public Class StarfieldDisplay
        Inherits BaseScreensaver
        Implements IScreensaver

        Private RandomDriver As Random
        Private CurrentWindowWidth As Integer
        Private CurrentWindowHeight As Integer
        Private ResizeSyncing As Boolean
        Private ReadOnly Stars As New List(Of Tuple(Of Integer, Integer))

        Public Overrides Property ScreensaverName As String = "Starfield" Implements IScreensaver.ScreensaverName

        Public Overrides Property ScreensaverSettings As Dictionary(Of String, Object) Implements IScreensaver.ScreensaverSettings

        Public Overrides Sub ScreensaverPreparation() Implements IScreensaver.ScreensaverPreparation
            'Variable preparations
            RandomDriver = New Random
            CurrentWindowWidth = Console.WindowWidth
            CurrentWindowHeight = Console.WindowHeight
            Console.BackgroundColor = ConsoleColor.Black
            Console.ForegroundColor = ConsoleColor.White
            Console.Clear()
        End Sub

        Public Overrides Sub ScreensaverLogic() Implements IScreensaver.ScreensaverLogic
            Console.CursorVisible = False

            'Move the stars left
            For Star As Integer = 0 To Stars.Count - 1
                Dim StarX As Integer = Stars(Star).Item1 - 1
                Dim StarY As Integer = Stars(Star).Item2
                Stars(Star) = New Tuple(Of Integer, Integer)(StarX, StarY)
            Next

            'If any star is out of X range, delete it
            For StarIndex As Integer = Stars.Count - 1 To 0 Step -1
                Dim Star = Stars(StarIndex)
                If Star.Item1 < 0 Then
                    'The star went beyond. Remove it.
                    Stars.RemoveAt(StarIndex)
                End If
            Next

            'Add new star if guaranteed
            Dim StarShowProbability As Double = 10 / 100
            Dim StarShowGuaranteed As Boolean = RandomDriver.NextDouble < StarShowProbability
            If StarShowGuaranteed Then
                Dim StarX As Integer = Console.WindowWidth - 1
                Dim StarY As Integer = RandomDriver.Next(Console.WindowHeight - 1)
                Stars.Add(New Tuple(Of Integer, Integer)(StarX, StarY))
            End If

            'Draw stars
            If CurrentWindowHeight <> Console.WindowHeight Or CurrentWindowWidth <> Console.WindowWidth Then ResizeSyncing = True
            If Not ResizeSyncing Then
                For StarIndex As Integer = Stars.Count - 1 To 0 Step -1
                    Dim Star = Stars(StarIndex)
                    Dim StarSymbol As Char = "*"c
                    Dim StarX As Integer = Star.Item1
                    Dim StarY As Integer = Star.Item2
                    WriteWhere(StarSymbol, StarX, StarY, False, ConsoleColor.White)
                Next
            Else
                WdbgConditional(ScreensaverDebug, DebugLevel.W, "Resize-syncing. Clearing...")
                Stars.Clear()
            End If

            'Reset resize sync
            ResizeSyncing = False
            CurrentWindowWidth = Console.WindowWidth
            CurrentWindowHeight = Console.WindowHeight
            SleepNoBlock(StarfieldDelay, ScreensaverDisplayerThread)
            Console.Clear()
        End Sub

    End Class
End Namespace

File: public/Kernel Simulator/Misc/Screensaver/Displays/Starfield.cs

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System;
using System.Collections.Generic;
using KS.Kernel.Debugging;
using KS.Misc.Threading;
using KS.Misc.Writers.ConsoleWriters;

namespace KS.Misc.Screensaver.Displays
{
    /// <summary>
    /// Settings for Starfield
    /// </summary>
    public static class StarfieldSettings
    {

        private static int _starfieldDelay = 10;

        /// <summary>
        /// [Starfield] How many milliseconds to wait before making the next write?
        /// </summary>
        public static int StarfieldDelay
        {
            get
            {
                return _starfieldDelay;
            }
            set
            {
                if (value <= 0)
                    value = 10;
                _starfieldDelay = value;
            }
        }

    }

    /// <summary>
    /// Display code for Starfield
    /// </summary>
    public class StarfieldDisplay : BaseScreensaver, IScreensaver
    {

        private Random RandomDriver;
        private int CurrentWindowWidth;
        private int CurrentWindowHeight;
        private bool ResizeSyncing;
        private readonly List<Tuple<int, int>> Stars = new();

        /// <inheritdoc/>
        public override string ScreensaverName { get; set; } = "Starfield";

        /// <inheritdoc/>
        public override Dictionary<string, object> ScreensaverSettings { get; set; }

        /// <inheritdoc/>
        public override void ScreensaverPreparation()
        {
            // Variable preparations
            RandomDriver = new Random();
            CurrentWindowWidth = ConsoleBase.ConsoleWrapper.WindowWidth;
            CurrentWindowHeight = ConsoleBase.ConsoleWrapper.WindowHeight;
            ConsoleBase.ConsoleWrapper.BackgroundColor = ConsoleColor.Black;
            ConsoleBase.ConsoleWrapper.ForegroundColor = ConsoleColor.White;
            ConsoleBase.ConsoleWrapper.Clear();
        }

        /// <inheritdoc/>
        public override void ScreensaverLogic()
        {
            ConsoleBase.ConsoleWrapper.CursorVisible = false;

            // Move the stars left
            for (int Star = 0, loopTo = Stars.Count - 1; Star <= loopTo; Star++)
            {
                int StarX = Stars[Star].Item1 - 1;
                int StarY = Stars[Star].Item2;
                Stars[Star] = new Tuple<int, int>(StarX, StarY);
            }

            // If any star is out of X range, delete it
            for (int StarIndex = Stars.Count - 1; StarIndex >= 0; StarIndex -= 1)
            {
                var Star = Stars[StarIndex];
                if (Star.Item1 < 0)
                {
                    // The star went beyond. Remove it.
                    Stars.RemoveAt(StarIndex);
                }
            }

            // Add new star if guaranteed
            double StarShowProbability = 10d / 100d;
            bool StarShowGuaranteed = RandomDriver.NextDouble() < StarShowProbability;
            if (StarShowGuaranteed)
            {
                int StarX = ConsoleBase.ConsoleWrapper.WindowWidth - 1;
                int StarY = RandomDriver.Next(ConsoleBase.ConsoleWrapper.WindowHeight - 1);
                Stars.Add(new Tuple<int, int>(StarX, StarY));
            }

            // Draw stars
            if (CurrentWindowHeight != ConsoleBase.ConsoleWrapper.WindowHeight | CurrentWindowWidth != ConsoleBase.ConsoleWrapper.WindowWidth)
                ResizeSyncing = true;
            if (!ResizeSyncing)
            {
                for (int StarIndex = Stars.Count - 1; StarIndex >= 0; StarIndex -= 1)
                {
                    var Star = Stars[StarIndex];
                    char StarSymbol = '*';
                    int StarX = Star.Item1;
                    int StarY = Star.Item2;
                    TextWriterWhereColor.WriteWhere(Convert.ToString(StarSymbol), StarX, StarY, false, ConsoleColor.White);
                }
            }
            else
            {
                DebugWriter.WriteDebugConditional(ref Screensaver.ScreensaverDebug, DebugLevel.W, "Resize-syncing. Clearing...");
                Stars.Clear();
            }

            // Reset resize sync
            ResizeSyncing = false;
            CurrentWindowWidth = ConsoleBase.ConsoleWrapper.WindowWidth;
            CurrentWindowHeight = ConsoleBase.ConsoleWrapper.WindowHeight;
            ThreadManager.SleepNoBlock(StarfieldSettings.StarfieldDelay, ScreensaverDisplayer.ScreensaverDisplayerThread);
            ConsoleBase.ConsoleWrapper.Clear();
        }

    }
}
